spark/src/main/scala/org/apache/spark/sql/comet/CometScanExec.scala

```diff
@@ -134,10 +135,42 @@ case class CometScanExec(
   override lazy val (outputPartitioning, outputOrdering): (Partitioning, Seq[SortOrder]) =
     (wrapped.outputPartitioning, wrapped.outputOrdering)
 
-  @transient
-  private lazy val pushedDownFilters = {
+  def translateToV1Filters(
+      dataFilters: Seq[Expression],
+      scalarSubqueryToLiteral: ScalarSubquery => Literal): Seq[Filter] = {
+    val scalarSubqueryReplaced = dataFilters.map(_.transform {
+      // Replace scalar subquery to literal so that `DataSourceStrategy.translateFilter` can
+      // support translating it.
+      case scalarSubquery: ScalarSubquery => scalarSubqueryToLiteral(scalarSubquery)
+    })
+
     val supportNestedPredicatePushdown = DataSourceUtils.supportNestedPredicatePushdown(relation)
-    dataFilters.flatMap(DataSourceStrategy.translateFilter(_, supportNestedPredicatePushdown))
+    // `dataFilters` should not include any constant metadata col filters
+    // because the metadata struct has been flatted in FileSourceStrategy
+    // and thus metadata col filters are invalid to be pushed down. Metadata that is generated
+    // during the scan can be used for filters.
+    scalarSubqueryReplaced
+      .filterNot(_.references.exists {
+        isFileSourceConstantMetadataAttribute
+      })
+      .flatMap(DataSourceStrategy.translateFilter(_, supportNestedPredicatePushdown))
+  }
+
+  @transient
+  private lazy val pushedDownFilters =
+    translateToV1Filters(dataFilters, q => convertScalarSubqueryToLiteral(q))
```

Comment:
Is it possible to define `pushedDownFilters` in Shims instead?
We can do the old way for Spark 3.x

For Spark 4.0, we can avoid the reflection like `convertScalarSubqueryToLiteral`